group_project_v2/stage/review.py

```diff
@@ -67,32 +87,47 @@ def validate(self):
 
         return violations
 
-    def _check_review_status(self, items_to_grade, review_items, review_item_key):
-        my_feedback = {
-            make_key(peer_review_item[review_item_key], peer_review_item["question"]): peer_review_item["answer"]
-            for peer_review_item in review_items
-            if peer_review_item['reviewer'] == self.anonymous_student_id
-        }
-
-        required_keys = [
-            make_key(item["id"], question.question_id)
-            for item in items_to_grade
+    def _group_review_items_by_reviewer(self, review_items):
+        empty_values = (None, '')
+        grouped = defaultdict(set)
+        for review_item in review_items:
+            if review_item["answer"] not in empty_values:  # exclude missing or empty answers
+                reviewer = review_item['reviewer']
+                review_item_key = make_key(review_item[self.review_item_key], review_item["question"])
+                grouped[reviewer].add(review_item_key)
+
+        return grouped
+
+    def _make_required_keys(self, items_to_grade):
+        return set(
+            make_key(item_id, question.question_id)
+            for item_id in items_to_grade
             for question in self.required_questions
-        ]
-        has_all = True
-        has_some = False
-
-        for key in required_keys:
-            has_answer = my_feedback.get(key, None) not in (None, '')
-            has_all = has_all and has_answer
-            has_some = has_some or has_answer
-
-        if has_all:
-            return ReviewState.COMPLETED
-        elif has_some:
-            return StageState.INCOMPLETE
-        else:
-            return StageState.NOT_STARTED
+        )
+
+    def _calculate_review_status(self, items_to_grade, reviewer_feedback_keys):
+        """
+        Calculates review status for all reviewers listed in review_items collection
+        :param collections.Iterable[int] items_to_grade: Ids of review subjects (teammates or other groups)
+        :param collections.Iterable[set[str]] reviewer_feedback_keys: Review feedback keys
+        :rtype: ReviewState
+        """
+        required_keys = self._make_required_keys(items_to_grade)
+        has_all = bool(required_keys) and reviewer_feedback_keys >= required_keys
+        has_some = bool(reviewer_feedback_keys & required_keys)
+
+        return self.REVIEW_STATE_CONDITIONS.get((has_some, has_all))
+
+    def _check_review_status(self, items_to_grade, review_items):
+        """
+        Calculates review status for current user
+        :param collections.Iterable[int] items_to_grade: Ids of review subjects (teammates or other groups)
+        :param collections.Iterable[dict] review_items: review items (answers to review questions)
+        :rtype: ReviewState
+        """
+        grouped_review_keys = self._group_review_items_by_reviewer(review_items)
```

Comment:
It seems inefficient to sort all the items by reviewer when you're just searching for one reviewer's items. Could you instead just iterate once over the full set, and pick out the items that match `self.anonymous_student_id` ?